arrow-libs/fx/arrow-fx-coroutines/src/commonMain/kotlin/arrow/fx/coroutines/Resource.kt

```diff
@@ -354,24 +351,15 @@ public fun <A> resource(block: suspend ResourceScope.() -> A): Resource<A> = blo
  * <!--- KNIT example-resource-06.kt -->
  */
 @ScopeDSL
-public suspend fun <A> resourceScope(action: suspend ResourceScope.() -> A): A {
-  val scope = ResourceScopeImpl()
-  val a: A = try {
-    action(scope)
-  } catch (e: Throwable) {
-    val ex = if (e is CancellationException) ExitCase.Cancelled(e) else ExitCase.Failure(e)
-    val ee = withContext(NonCancellable) {
-      scope.cancelAll(ex, e) ?: e
-    }
-    throw ee
-  }
-  withContext(NonCancellable) {
-    scope.cancelAll(ExitCase.Completed)?.let { throw it }
-  }
-  return a
+@OptIn(DelicateCoroutinesApi::class)
+@Suppress("REDUNDANT_INLINE_SUSPEND_FUNCTION_TYPE")
+public suspend inline fun <A> resourceScope(action: suspend ResourceScope.() -> A): A {
+  val (scope, cancelAll) = resource { this }.allocated()
```

Comment:
A little hacky, but it results in no new ABI added. We might instead want a function `allocateResourceScope(): Pair<ResourceScope, suspend (ExitCase) -> Unit>`

Reply:
I think this is fine - if anything a comment to highlight that `this` is the `ResourceScope` for clarity but it's internal and imo understandable from the context.